src/tools/erasure-code/ceph-erasure-code-tool.cc

```diff
@@ -0,0 +1,253 @@
+  for (auto &opt_str : profile_str) {
+    std::vector<std::string> opt;
+    boost::split(opt, opt_str, boost::is_any_of("="));
+    if (opt.size() <= 1) {
+      usage("invalid profile", std::cerr);
+      return 1;
+    }
+    profile[opt[0]] = opt[1];
+  }
+  auto plugin = profile.find("plugin");
+  if (plugin == profile.end()) {
+      usage("invalid profile: plugin not specified", std::cerr);
+      return 1;
+  }
+
+  ceph::ErasureCodeInterfaceRef ec_impl;
+  stringstream ss;
+  ceph::ErasureCodePluginRegistry::instance().factory(
+    plugin->second, g_conf().get_val<std::string>("erasure_code_dir"),
+    profile, &ec_impl, &ss);
+  if (!ec_impl) {
+    usage("invalid profile: " + ss.str(), std::cerr);
+    return 1;
+  }
+
+  if (command == "validate-profile") {
+    if (args.size() > 2) {
+      std::set<std::string> valid_params(display_params.begin(),
+                                         display_params.end());
+      display_params.clear();
+      for (size_t i = 2; i < args.size(); i++) {
+        if (!valid_params.count(args[i])) {
+          usage("invalid display param: " + std::string(args[i]), std::cerr);
+          return 1;
+        }
+        display_params.push_back(args[i]);
+      }
+    }
+
+    for (auto &param : display_params) {
+      if (display_params.size() > 1) {
+        std::cout << param << ": ";
+      }
+      if (param == "chunk_count") {
+        std::cout << ec_impl->get_chunk_count() << std::endl;
+      } else if (param == "data_chunk_count") {
+        std::cout << ec_impl->get_data_chunk_count() << std::endl;
+      } else if (param == "coding_chunk_count") {
+        std::cout << ec_impl->get_coding_chunk_count() << std::endl;
+      } else {
+        ceph_abort_msgf("unknown display_param: %s", param.c_str());
+      }
+    }
+
+    return 0;
+  }
+
+  if (command == "calc-chunk-size") {
+    if (args.size() < 3) {
+      usage("not enought arguments", std::cerr);
+      return 1;
+    }
+
+    uint64_t object_size = atoi(args[2]);
+    if (object_size <= 0) {
+      usage("invalid object size", std::cerr);
+      return 1;
+    }
+
+    std::cout << ec_impl->get_chunk_size(object_size) << std::endl;
+    return 0;
+  }
+
+  if (args.size() < 5) {
+    usage("not enought arguments", std::cerr);
+    return 1;
+  }
+
+  uint64_t stripe_unit = atoi(args[2]);
+  if (stripe_unit <= 0) {
+    usage("invalid stripe unit", std::cerr);
+    return 1;
+  }
+
+  uint64_t stripe_size = atoi(profile["k"].c_str());
+  ceph_assert(stripe_size > 0);
+  uint64_t stripe_width = stripe_size * stripe_unit;
+  ECUtil::stripe_info_t sinfo(stripe_size, stripe_width);
+
+  std::map<int, ceph::bufferlist> encoded_data;
+  std::vector<std::string> shards;
+  boost::split(shards, args[3], boost::is_any_of(","));
+  for (auto &shard : shards) {
+    encoded_data[atoi(shard.c_str())] = {};
+  }
+  ceph::bufferlist decoded_data;
+  std::string fname = args[4];
+
+  if (command == "encode") {
```

Comment:
better off extracting the code handling commands into their own functions for better readability.